File: sdk/kotlin/src/test/kotlin/com/zseven/zode/ProtocolTest.kt
package com.zseven.zode

import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class ProtocolTest {
    @Test
    fun protocolMethodsMatchSchemaInOrder() {
        val schemaRaw = fixturesDir().resolve("protocol.schema.json").readText()
        val schema = testJson.parseToJsonElement(schemaRaw) as JsonObject
        val methods =
            (schema["methods"] as kotlinx.serialization.json.JsonArray).map {
                testJson.decodeFromJsonElement(String.serializer(), it)
            }

        assertEquals(27, ProtocolMethod.entries.size, "SDK must expose 27 methods")
        assertEquals(methods.size, ProtocolMethod.entries.size, "schema/SDK method count differs")
        methods.forEachIndexed { i, want ->
            assertEquals(want, ProtocolMethod.entries[i].wireName, "method $i mismatch")
        }
    }

    @Test
    fun classifierTagsEachKind() {
        assertEquals(
            FrameKind.Response,
            classifyIncomingFrame("""{"jsonrpc":"2.0","id":1,"result":{}}""", testJson).kind,
        )
        assertEquals(
            FrameKind.Error,
            classifyIncomingFrame(
                """{"jsonrpc":"2.0","id":1,"error":{"code":-1,"message":"x"}}""",
                testJson,
            ).kind,
        )
        assertEquals(
            FrameKind.Notification,
            classifyIncomingFrame("""{"jsonrpc":"2.0","method":"turn/started","params":{}}""", testJson).kind,
        )
        assertEquals(
            FrameKind.ServerRequest,
            classifyIncomingFrame(
                """{"jsonrpc":"2.0","id":"a","method":"approval/request","params":{}}""",
                testJson,
            ).kind,
        )
    }

    @Test
    fun classifierRejectsFramesMissingJsonrpc() {
        val bad =
            listOf(
                """{"id":1,"result":{}}""",
                """{"jsonrpc":"1.0","id":1,"result":{}}""",
                """{"jsonrpc":"2.0","id":1}""",
                "not json",
            )
        for (raw in bad) {
            assertFailsWith<InvalidFrameException>("expected rejection for $raw") {
                classifyIncomingFrame(raw, testJson)
            }
        }
    }

    @Test
    fun initializeParamsOmitsNullApprovalPolicy() {
        val omitted =
            testJson.encodeToJsonElement(
                InitializeParams.serializer(),
                InitializeParams(ClientInfo("n", "v"), null),
            ) as JsonObject
        assertFalse(omitted.containsKey("approvalPolicy"), "null approvalPolicy must be omitted")

        val present =
            testJson.encodeToJsonElement(
                InitializeParams.serializer(),
                InitializeParams(ClientInfo("n", "v"), "auto"),
            ) as JsonObject
        assertTrue(present.containsKey("approvalPolicy"), "explicit approvalPolicy must be present")
    }

    @Test
    fun requestFrameAlwaysCarriesJsonrpc() {
        val frame =
            JsonRpcRequest(
                id = kotlinx.serialization.json.JsonPrimitive(1),
                method = "initialize",
                params = buildJsonObject { put("k", "v") },
            )
        val obj = testJson.encodeToJsonElement(JsonRpcRequest.serializer(), frame) as JsonObject
        assertEquals("2.0", (obj["jsonrpc"] as kotlinx.serialization.json.JsonPrimitive).content)
    }
}

File: sdk/go/protocol_test.go
package zodesdk

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"testing"
)

// schemaPath resolves fixtures/jsonrpc/protocol.schema.json relative to this
// test file so the test does not depend on the working directory.
func fixturesDir(t *testing.T) string {
	t.Helper()
	_, thisFile, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	// sdk/go/protocol_test.go -> sdk/fixtures/jsonrpc
	return filepath.Join(filepath.Dir(thisFile), "..", "fixtures", "jsonrpc")
}

func TestProtocolMethodsMatchSchema(t *testing.T) {
	raw, err := os.ReadFile(filepath.Join(fixturesDir(t), "protocol.schema.json"))
	if err != nil {
		t.Fatalf("read schema: %v", err)
	}
	var schema struct {
		Methods []string `json:"methods"`
	}
	if err := json.Unmarshal(raw, &schema); err != nil {
		t.Fatalf("parse schema: %v", err)
	}
	if len(ProtocolMethods) != 27 {
		t.Fatalf("expected 27 methods, got %d", len(ProtocolMethods))
	}
	if len(schema.Methods) != len(ProtocolMethods) {
		t.Fatalf("schema has %d methods, SDK has %d", len(schema.Methods), len(ProtocolMethods))
	}
	for i, want := range schema.Methods {
		if got := ProtocolMethods[i].String(); got != want {
			t.Fatalf("method %d: schema %q, SDK %q", i, want, got)
		}
	}
}

func TestClassifyIncomingFrameKinds(t *testing.T) {
	cases := []struct {
		raw  string
		want FrameKind
	}{
		{`{"jsonrpc":"2.0","id":1,"result":{}}`, FrameResponse},
		{`{"jsonrpc":"2.0","id":1,"error":{"code":-1,"message":"x"}}`, FrameError},
		{`{"jsonrpc":"2.0","method":"turn/started","params":{}}`, FrameNotification},
		{`{"jsonrpc":"2.0","id":"a","method":"approval/request","params":{}}`, FrameServerRequest},
	}
	for _, tc := range cases {
		frame, err := ClassifyIncomingFrame([]byte(tc.raw))
		if err != nil {
			t.Fatalf("classify %s: unexpected error %v", tc.raw, err)
		}
		if frame.Kind != tc.want {
			t.Fatalf("classify %s: got %q want %q", tc.raw, frame.Kind, tc.want)
		}
	}
}

func TestClassifyRejectsFramesMissingJSONRPC(t *testing.T) {
	bad := []string{
		`{"id":1,"result":{}}`,
		`{"jsonrpc":"1.0","id":1,"result":{}}`,
		`{"jsonrpc":"2.0","id":1}`,
		`not json`,
	}
	for _, raw := range bad {
		if _, err := ClassifyIncomingFrame([]byte(raw)); err == nil {
			t.Fatalf("expected rejection for %q", raw)
		}
	}
}

func TestInitializeParamsOmitsEmptyApprovalPolicy(t *testing.T) {
	built, err := json.Marshal(NewInitializeParams("n", "v", ""))
	if err != nil {
		t.Fatal(err)
	}
	var fields map[string]json.RawMessage
	if err := json.Unmarshal(built, &fields); err != nil {
		t.Fatal(err)
	}
	if _, ok := fields["approvalPolicy"]; ok {
		t.Fatalf("empty approvalPolicy must be omitted, got %s", built)
	}
	built2, _ := json.Marshal(NewInitializeParams("n", "v", "auto"))
	if err := json.Unmarshal(built2, &fields); err != nil {
		t.Fatal(err)
	}
	if _, ok := fields["approvalPolicy"]; !ok {
		t.Fatalf("explicit approvalPolicy must be present, got %s", built2)
	}
}
